src/app/modules/spcp/sp.oidc.client.ts

```diff
@@ -0,0 +1,583 @@
+import axios from 'axios'
+import { ObjectId } from 'bson'
+import { createPrivateKey, createPublicKey, KeyObject } from 'crypto'
+import {
+  compactDecrypt,
+  JWTPayload,
+  jwtVerify,
+  JWTVerifyResult,
+  SignJWT,
+} from 'jose'
+import jwkToPem from 'jwk-to-pem'
+import NodeCache from 'node-cache'
+import { BaseClient, Issuer } from 'openid-client'
+import promiseRetry from 'promise-retry'
+
+import {
+  CreateAuthorisationUrlError,
+  CreateJwtError,
+  ExchangeAuthTokenError,
+  GetDecryptionKeyError,
+  GetVerificationKeyError,
+  InvalidIdTokenError,
+  JwkError,
+  MissingIdTokenError,
+  VerificationKeyError,
+} from './sp.oidc.client.errors'
+import {
+  CryptoKeySet,
+  isEC,
+  isECPrivate,
+  isSigningKey,
+  PublicJwks,
+  Refresh,
+  SecretJwks,
+  SigningKey,
+  SpOidcClientCacheConstructorParams,
+  SpOidcClientConstructorParams,
+} from './sp.oidc.client.types'
+
+/**
+ * Cache class which provides read-through capability and refresh-ahead before expiry
+ * Handles discovery and retrieval of NDI's public jwks
+ * Exported for testing
+ */
+export class SpOidcClientCache extends NodeCache {
+  #spOidcNdiDiscoveryEndpoint: string
+  #spOidcNdiJwksEndpoint: string
+  #spOidcRpClientId: string
+  #spOidcRpRedirectUrl: string
+  #spOidcRpSecretJwks: SecretJwks
+
+  /**
+   * Constructor for cache
+   * On instantiation, trigger a refresh to populate the cache
+   */
+  constructor({
+    options,
+    spOidcNdiDiscoveryEndpoint,
+    spOidcNdiJwksEndpoint,
+    spOidcRpClientId,
+    spOidcRpRedirectUrl,
+    spOidcRpSecretJwks,
+  }: SpOidcClientCacheConstructorParams) {
+    super(options)
+
+    this.#spOidcNdiDiscoveryEndpoint = spOidcNdiDiscoveryEndpoint
+    this.#spOidcNdiJwksEndpoint = spOidcNdiJwksEndpoint
+    this.#spOidcRpClientId = spOidcRpClientId
+    this.#spOidcRpRedirectUrl = spOidcRpRedirectUrl
+    this.#spOidcRpSecretJwks = spOidcRpSecretJwks
+
+    // On expiry, refresh cache. If fail to refresh, log but do not throw error.
+    this.on('expired', () =>
+      this.refresh().catch((err) =>
+        console.warn(
```

Comment:
No logger for this?

Reply:
No for now, because the intent was to keep the client standalone as much as possible - see [discussion](https://github.com/opengovsg/FormSG/pull/4011#discussion_r902153742), but welcome thoughts on it